### PR TITLE
Add Luigi

### PR DESCRIPTION
**Pull request by DrKwint**

### CODE REVIEW

File: luigi.cfg [hunk 1]
```diff
@@ -0,0 +1,8 @@
+[core]
+default-scheduler-host = pyroclast-luigi.gilgi.org
+default-scheduler-port = 80
```
Comment (ihowell) [inline luigi.cfg]: Wrong. Also you don't need to set it. You only need to set default-scheduler-url. Luigi will take care of everything else.

File: pyroclast/common/feature_classifier.py [hunk 2]
```diff
@@ -115,13 +123,11 @@ def input_search(self, x, feature_target, search_method='fgm'):
...
             max_iters = 100
             forward_fn = lambda _x: tf.norm(feature_target - self.features(_x),
-                                            2)  # + 0.1 * tf.norm(x - _x, 2)
-            for i in range(1, max_iters + 1):
+                                            2)  # + 1e-6 * tf.norm(x - _x, 2)
+            for i in range(max_iters):
                 print('forward_fn', forward_fn(x))
                 delta = fast_gradient_method(forward_fn, x, 0.001, 2)
                 x += delta
-                if forward_fn(x) < 1:
```
Comment (ihowell) [inline pyroclast/common/feature_classifier.py]: How do you break out early now?

File: pyroclast/luigi/pyro_tasks.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import luigi
+import functools
+import itertools
+import json
+import random
+
+from pyroclast.luigi.train_task import TrainTask
+
+
+class TopTask(luigi.Task):
```
Comment (ihowell) [inline pyroclast/luigi/pyro_tasks.py]: Huh? What does this do?

Comment (ihowell) [reply pyroclast/luigi/pyro_tasks.py]: Also, it looks like it's just running other tasks, so it should extend `luigi.WrapperTask`.  Further, we don't actually need any top level task, unless you want your pretty trees in the visualizer.

File: pyroclast/luigi/resources/cpu.sh [hunk 4]
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
```
Comment (ihowell) [inline pyroclast/luigi/resources/cpu.sh]: Rename to `crane_cpu.sh`. Same with gpu. Then have another resource that is `local_cpu`. No need for preambles in the luigi directory.

File: pyroclast/luigi/train_task.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    batch_size = luigi.IntParameter()
+    seed = luigi.IntParameter()
+
+    learning_rate = luigi.FloatParameter()
+    conv_stack_name = luigi.Parameter()
+    lambd = luigi.FloatParameter()
+
+    def get_task_name_str(self):
+        id_str = 'learn'
+        id_str += '_{}'.format(self.dataset)
```
Comment (ihowell) [inline pyroclast/luigi/train_task.py]: Umm, how about no? Create an array and join with `'_'.join(...)`

File: luigi.cfg (hunk 1 above)
Comment (DrKwint) [reply luigi.cfg]: Oh, gilgi gave me this from his config. Is it bad enough that we need to delete it?

File: pyroclast/luigi/pyro_tasks.py (hunk 3 above)
Comment (DrKwint) [reply pyroclast/luigi/pyro_tasks.py]: I'll look into that, ty

File: pyroclast/luigi/train_task.py (hunk 5 above)
Comment (DrKwint) [reply pyroclast/luigi/train_task.py]: This method has the virtue of being very flexible to change and update. Is that not reason enough to leave it like this?

File: luigi.cfg (hunk 1 above)
Comment (ihowell) [reply luigi.cfg:3]: Yes, they could potentially conflict later on. KISS (keep it simple stupid)

File: pyroclast/features/features.py [hunk 6]
```diff
@@ -273,37 +273,38 @@ def plot_input_grads(data_dict,
     args_template = args
 
     models = {}
-    if not os.path.exists('mnist_normal'):
-        model_name = 'mnist_normal'
-        models[model_name] = learn(**args)
-
-    for lambd in [1e0, 1e1, 1e2]:
-        for alpha in [0., 1e0, 1e1, 1e2]:
+    for lambd in [0., 1e0, 1e1, 1e2]:
```
Comment (ihowell) [inline pyroclast/features/features.py:276]: Why not just use an itertools.product here? If we continue like this, every time we need to vary and stop varying another parameter, then the entire next section will have to change indentation.

Comment (DrKwint) [reply pyroclast/features/features.py:276]: The whole point of adding Luigi is to deal with this problem. This code is moot as soon as Luigi is securely in place.
